### PR TITLE
CR feedback and unit tests for MQTT modules.

### PR DESCRIPTION
**Pull request by BertKleewein**
This is larger than I thought it would be.  

I did the following:
1) Removed the onInputMessage() and removeInputMessageHandler() functions, opting for on('inputMessage', function(inputName, msg)) instead.
2) Added enableInputMessages and disableInputMessages at the transport level.
3) Diverged code paths for sendEvent and sendOutputEvent.  A module can do both Event and OutputEvent.  Likewise, a device can do both.  
4) Diverged code paths for on('message') and on('inputMessage').
5) Added unit test coverage.

I have mixed feelings about this change.  It feels like I've added a lot of complexity that maybe isn't necessary.

### CODE REVIEW

File: common/transport/mqtt/devdoc/mqtt_base_requirements.md [hunk 1]
```diff
@@ -38,7 +38,7 @@ The `Mqtt` constructor receives the configuration parameters to configure the MQ
 
 ### MqttBase.connect(config, done)
 The `connect` method establishes a connection with the server using the config object passed in the arguments.
-**SRS_NODE_COMMON_MQTT_BASE_16_006: [** The `connect` method shall throw a ReferenceError if the config argument is falsy, or if one of the following properties of the config argument is falsy: deviceId, host, and one of sharedAccessSignature or x509.cert and x509.key. **]**
+**SRS_NODE_COMMON_MQTT_BASE_16_006: [** The `connect` method shall throw a ReferenceError if the config argument is falsy, or if one of the following properties of the config argument is falsy: uri, clientId, username, and one of sharedAccessSignature or x509.cert and x509.key. **]**
```
Comment (pierreca) [inline common/transport/mqtt/devdoc/mqtt_base_requirements.md]: >throw [](start = 67, length = 5)

change is good but as a general remark, since these APIs are internal and we use typescript now, I would rather see strong typing than runtime checks in the future. Don't be surprised if you see me removing them in my PRs :)

File: common/transport/mqtt/test/_mqtt_base_test.js [hunk 2]
```diff
@@ -14,9 +14,24 @@ var Message = require('azure-iot-common').Message;
 var endpoint = require('azure-iot-common').endpoint;
 var errors = require('azure-iot-common').errors;
 
+
+
+
+
```
Comment (pierreca) [inline common/transport/mqtt/test/_mqtt_base_test.js]: ok super nitpicky but watch out for unnecessary whitespace? I know, i'm nitpicky. sorry.

File: common/transport/mqtt/test/_mqtt_base_test.js [hunk 3]
```diff
@@ -26,31 +41,15 @@ describe('MqttBase', function () {
...
-          host: "host.name",
-          deviceId: deviceId,
-          sharedAccessSignature: "sasToken"
-        };
-        var mqtt = new MqttBase(config, 'test', new FakeMqtt());
-        assert.throws(function () {
-          mqtt.connect(config, function () {});
-        }, ReferenceError, 'Invalid transport configuration');
+    ['uri', 'clientId', 'username'].forEach(function(falsyProperty) {
+      [null, undefined].forEach(function (falsyValue) {
```
Comment (pierreca) [inline common/transport/mqtt/test/_mqtt_base_test.js]: >undefined [](start = 13, length = 9)

empty string too? ''?

File: device/core/src/client.ts [hunk 4]
```diff
@@ -65,6 +65,7 @@ export class Client extends EventEmitter {
   private blobUploadClient: BlobUploadClient; // TODO: wrong casing/naming convention
   private _c2dEnabled: boolean;
   private _methodsEnabled: boolean;
+  private _inputMessagesEnabled: boolean;
```
Comment (pierreca) [inline device/core/src/client.ts]: >_inputMessagesEnabled [](start = 10, length = 21)

note for later: there's definitely a pattern emerging here. wondering if the BehavioralFsm would help.

File: device/transport/mqtt/devdoc/mqtt_requirements.md [hunk 5]
```diff
@@ -348,3 +417,22 @@ interface MethodMessage {
 ```
 **]**
 
+### inputMessage event
+
+**SRS_NODE_DEVICE_MQTT_18_057: [** If there is a listener for the `inputMessage` event, a `inputMessage` event shall be emitted for each message received. **]**
```
Comment (pierreca) [inline device/transport/mqtt/devdoc/mqtt_requirements.md]: >If there is a listener for the `inputMessage` event, [](start = 35, length = 52)

you can emit even if there are no listeners, it is not going to crash. not sure if there's value if that first "if" statement since we're using enable/disable anyway?

File: device/transport/mqtt/test/_mqtt_receiver_test.js [hunk 6]
```diff
@@ -329,5 +330,24 @@ describe('Mqtt as MqttReceiver', function () {
...
+    /*Tests_SRS_NODE_DEVICE_MQTT_18_058: [ When an `inputMessage` event is received, Mqtt shall extract the inputName from the topic according to the following convention: 'devices/<deviceId>/modules/<moduleId>/inputs/<inputName>' ]*/
+    /*Tests_SRS_NODE_DEVICE_MQTT_18_056: [ When an `inputMessage` event is emitted, the first parameter shall be the inputName and the second parameter shall be of type `Message`. ]*/
+    describe('#inputMessage', function() {
+        it('emits an inputMessage event with a Message object when there is a listener', function (done) {
+          fakeConfig.moduleId = 'moduleId';
+          var receiver = new Mqtt(fakeAuthenticationProvider, fakeMqttBase);
+          receiver.connect(function () {
+            receiver.on('inputMessage', function (inputName, msg) {
+              assert.strictEqual(inputName, 'fakeInputName')
+              assert.instanceOf(msg, Message);
```
Comment (pierreca) [inline device/transport/mqtt/test/_mqtt_receiver_test.js]: >assert.instanceOf(msg, Message); [](start = 14, length = 32)

is that working? you're not passing a message in the emit?

File: device/transport/mqtt/test/_mqtt_test.js [hunk 7]
```diff
@@ -794,3 +923,9 @@ describe('Mqtt', function () {
     });
   });
 });
+
+
+
+
+
+
```
Comment (pierreca) [inline device/transport/mqtt/test/_mqtt_test.js]: nit: whitespaaace :-/

File: device/transport/mqtt/test/_mqtt_receiver_test.js (hunk 6 above)
Comment (BertKleewein) [reply device/transport/mqtt/test/_mqtt_receiver_test.js]: Yep, this was copy/paste from the corresponding message test

---
In reply to: [171429827](https://github.com/Azure/azure-iot-sdk-node/pull/237#discussion_r171429827) [](ancestors = 171429827)
